librz/util/print.c

```diff
@@ -2074,40 +2074,59 @@ RZ_API int rz_print_jsondump(RzPrint *p, const ut8 *buf, int len, int wordsize)
 	return words;
 }
 
-RZ_API void rz_print_hex_from_bin(RzPrint *p, char *bin_str) {
-	int i, j, index;
-	RzPrint myp = { .cb_printf = libc_printf };
-	const int len = strlen(bin_str);
-	if (!len) {
+RZ_API void rz_print_hex_from_base2(RzPrint *p, char *base2) {
+	bool first = true;
+	const int len = strlen(base2);
+	if (len < 1) {
 		return;
 	}
-	ut64 n, *buf = malloc(sizeof(ut64) * ((len + 63) / 64));
-	if (!buf) {
-		eprintf("allocation failed\n");
+
+	RzPrint defprint = { .cb_printf = libc_printf };
+	if (!p) {
+		p = &defprint;
+	}
+
+	// we split each section by 8 bits and have bytes.
+	ut32 bytes_size = (len >> 3) + (len & 7 ? 1 : 0);
+	ut8 *bytes = calloc(bytes_size, sizeof(ut8));
+	if (!bytes) {
+		eprintf("cannot allocate %d bytes\n", bytes_size);
 		return;
 	}
-	if (!p) {
-		p = &myp;
+
+	int c = len & 7;
+	if (c) {
+		// align counter to 8 bits
+		c = 8 - c;
 	}
-	for (i = len - 1, index = 0; i >= 0; i -= 64, index++) {
-		n = 0;
-		for (j = 0; j < 64 && i - j >= 0; j++) {
-			n += (ut64)(bin_str[i - j] - '0') << j;
+	for (int i = 0, j = 0; i < len && j < bytes_size; i++, c++) {
+		if (base2[i] != '1' && base2[i] != '0') {
+			eprintf("invalid base2 number %c at char %d\n", base2[i], i);
+			free(bytes);
+			return;
 		}
-		buf[index] = n;
+		// c & 7 is c % 8
+		if (c > 0 && !(c & 7)) {
+			j++;
+		}
+		bytes[j] <<= 1;
+		bytes[j] |= base2[i] - '0';
 	}
-	index--;
+
 	p->cb_printf("0x");
-	while (index > 0 && !buf[index]) {
-		index--;
-	}
-	p->cb_printf("%" PFMT64x, buf[index]);
-	index--;
-	for (i = index; i >= 0; i--) {
-		p->cb_printf("%016" PFMT64x, buf[i]);
+	for (int i = 0; i < bytes_size; ++i) {
+		if (first) {
+			if (i != (bytes_size - 1) && !bytes[i]) {
+				continue;
+			}
+			p->cb_printf("%x", bytes[i]);
```

Comment:
Unrelated to this PR, but probably we want to put everything that RzPrint does in the RzStrBuf, then print it from the RzCore.